File: src/main/java/no/fintlabs/consumer/exception/kafka/ConsumerErrorPublisher.java
package no.fintlabs.consumer.exception.kafka;

import lombok.extern.slf4j.Slf4j;
import no.novari.kafka.producing.ParameterizedProducerRecord;
import no.novari.kafka.producing.ParameterizedTemplate;
import no.novari.kafka.producing.ParameterizedTemplateFactory;
import no.novari.kafka.topic.EventTopicService;
import no.novari.kafka.topic.configuration.EventCleanupFrequency;
import no.novari.kafka.topic.configuration.EventTopicConfiguration;
import no.novari.kafka.topic.name.EventTopicNameParameters;
import no.novari.kafka.topic.name.TopicNamePrefixParameters;
import no.fintlabs.status.models.error.ConsumerError;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.util.UUID;

@Slf4j
@Service
public class ConsumerErrorPublisher {

    private final ParameterizedTemplate<ConsumerError> eventProducer;
    private final EventTopicNameParameters eventName;
    private static final Duration RETENTION_TIME = Duration.ofDays(7);
    private static final int PARTITIONS = 1;

    public ConsumerErrorPublisher(
            ParameterizedTemplateFactory parameterizedTemplateFactory,
            EventTopicService eventTopicService
    ) {
        this.eventProducer = parameterizedTemplateFactory.createTemplate(ConsumerError.class);
        this.eventName = createEventName();
        eventTopicService.createOrModifyTopic(
                eventName,
                EventTopicConfiguration
                        .stepBuilder()
                        .partitions(PARTITIONS)
                        .retentionTime(RETENTION_TIME)
                        .cleanupFrequency(EventCleanupFrequency.NORMAL)
                        .build()
        );
    }

    public void publish(ConsumerError consumerError) {
        log.info("Publishing consumer-error to Kafka!");
        eventProducer.send(
                ParameterizedProducerRecord.<ConsumerError>builder()
                        .key(UUID.randomUUID().toString())
                        .topicNameParameters(eventName)
                        .value(consumerError)
                        .build()
        );
    }

    private EventTopicNameParameters createEventName() {
        return EventTopicNameParameters.builder()
                .topicNamePrefixParameters(
                        TopicNamePrefixParameters
                                .stepBuilder()
                                .orgId("fintlabs-no")
                                .domainContextApplicationDefault()
                                .build()
                )
                .eventName("consumer-error")
                .build();
    }

}

File: src/main/java/no/fintlabs/consumer/exception/kafka/ConsumerErrorPublisher.kt
package no.fintlabs.consumer.exception.kafka

import no.fintlabs.consumer.config.OrgId
import no.fintlabs.status.models.error.ConsumerError
import no.novari.kafka.producing.ParameterizedProducerRecord
import no.novari.kafka.producing.ParameterizedTemplate
import no.novari.kafka.producing.ParameterizedTemplateFactory
import no.novari.kafka.topic.EventTopicService
import no.novari.kafka.topic.configuration.EventCleanupFrequency
import no.novari.kafka.topic.configuration.EventTopicConfiguration
import no.novari.kafka.topic.name.EventTopicNameParameters
import no.novari.kafka.topic.name.TopicNamePrefixParameters
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.Duration
import java.util.UUID

@Service
class ConsumerErrorPublisher(
    parameterizedTemplateFactory: ParameterizedTemplateFactory,
    eventTopicService: EventTopicService,
) {
    private val logger = LoggerFactory.getLogger(javaClass)
    private val eventProducer: ParameterizedTemplate<ConsumerError> =
        parameterizedTemplateFactory.createTemplate(ConsumerError::class.java)
    private val eventName: EventTopicNameParameters = createEventName()

    init {
        eventTopicService.createOrModifyTopic(
            eventName,
            EventTopicConfiguration
                .stepBuilder()
                .partitions(PARTITIONS)
                .retentionTime(RETENTION_TIME)
                .cleanupFrequency(EventCleanupFrequency.NORMAL)
                .build(),
        )
    }

    fun publish(consumerError: ConsumerError) {
        logger.info("Publishing consumer-error to Kafka!")
        eventProducer.send(
            ParameterizedProducerRecord
                .builder<ConsumerError>()
                .key(UUID.randomUUID().toString())
                .topicNameParameters(eventName)
                .value(consumerError)
                .build(),
        )
    }

    private fun createEventName(): EventTopicNameParameters =
        EventTopicNameParameters
            .builder()
            .topicNamePrefixParameters(
                TopicNamePrefixParameters
                    .stepBuilder()
                    .orgId(FINTLABS_ORG_ID.asTopicSegment)
                    .domainContextApplicationDefault()
                    .build(),
            ).eventName("consumer-error")
            .build()

    companion object {
        private val FINTLABS_ORG_ID = OrgId.from("fintlabs.no")
        private val RETENTION_TIME: Duration = Duration.ofDays(7)
        private const val PARTITIONS = 1
    }
}
